cli: show the current database name in the prompt.

knz opened the pull request:
This completes #12747 by adding the missing test.

It also fixes the code to avoid retrieving the db state while a transaction is ongoing and not in a "simple" state.
It also fixes a couple of broken tests in `test_txn_prompts.tcl` that were introduced in 282b76ccebf450613e634ea7df5e768c931a4ea1.

Fixes #12038.

cc @pmamatsis.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13379)
<!-- Reviewable:end -->

pmamatsis committed:
cli: show the current database name in the prompt.

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/sql.go, line 457 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13379#-KcKJGxaOdIZptqzvpfu:-KcKJGxaOdIZptqzvpfv:b-n51r7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/34ccf3e5aeca6d223066297c1fe5ab3082f808f6/pkg/cli/sql.go#L457)):*
> ```Go
> 
> 	query := makeQuery(`SHOW DATABASE`)
> 	rows, err := query(c.conn)
> ```

`rows` does not appear to be closed anywhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13379)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/sql.go, line 457 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13379#-KcKJGxaOdIZptqzvpfu:-KcNOKOrb6fWnilTJ0gO:b-c9cdis) ([raw file](https://github.com/cockroachdb/cockroach/blob/34ccf3e5aeca6d223066297c1fe5ab3082f808f6/pkg/cli/sql.go#L457)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

`rows` does not appear to be closed anywhere.
</blockquote></details>

Wow, good find.
This means the earlier PR #10866  also introduced a leak in refreshTransactionStatus below. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13379)*
<!-- Sent from Reviewable.io -->

pmamatsis commented:
Hi @knz and @mjibson,
   do i need to add this in my sql.go file?

Best regards,
Panos.

knz commented:
@pmamatsis once we merge this PR you can rebase and you'll get the change automatically.

But yes otherwise do not forget to Close the results of query().

pmamatsis commented:
Hi @knz,
   i will wait for the merge then...but since i am going to work on the issue : #12408 is it possible to give me some tips on rebasing? I have tried to do it the last time and ended up breaking everything. Actually i have just followed these instructons:

https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Best regards and thank you in advance,
Panos.

knz commented:
@pmamatsis rebases are pretty important so it's good that you want to make them work :)

So the way I do it is like this.

Suppose I am working on some branch **my-feature-branch**. 
At some point I want to pick up the latest changes from **master**.
I do this like this:

1. commit my current work on **my-feature-branch**
2. change to master: `git checkout master`
3. forward to the latest version: `git pull` (or `git pull origin master` if you have multiple origins)
4. change back to my branch: `git checkout my-feature-branch`
5. rebase: `git rebase master`

At this point, if there is no conflict, then it's all I have to do.  A simple `git push -f` is sufficient to update my own branch on github for a pull request.

However sometimes step 5 causes *merge conflicts*. In this case:

6. edit the files that have caused a conflict. Search for regions with the form `<<<`  ... `===` ... `>>>` and fix the conflict.
7. mark the conflict as resolved with: `git add your-fixed-file.go`
8. complete the rebase with `git rebase --continue`

If you cannot fix the conflict at step 6, then you can abort the rebase entirely (and get back to your original code) with `git rebase --abort`.


Now, remember, for all this to work, you must **never merge the master branch into your branch** (with e.g. `git pull master` or `git merge master` while you have your branch checked out). Rebasing only works properly if your branch is entirely forked off a single **master** commit.

I hope this helps.

knz commented:
Note that my explanation above is only about rebasing.

You also need to squash commits, but that is a different thing, so I did not want to make the explanation too complex. I think you should delay learning about squashes until you understand how to rebase successfully.

pmamatsis commented:
Hi @knz,
   really a big thank you for your helpful guide. I will try to do my rebase tomorrow evening (local Greece time) in order to have a much more clearer head! Tonight though i will spend it studying your guide! :)

Best regards,
Panos.

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13379#-:-KcOwjK24fuvu3qEaeDO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR

pmamatsis commented:
Good evening @knz,
   can i start working on the issue "cli: missing username in connection URL prevents username from appearing in prompt" ? I will find the appropriate thread from the list of issues and continue my feedback there.

Best regards,
Panos.

knz commented:
Excellent, thanks

pmamatsis <notifications@github.com> schreef op 10 februari 2017 19:52:01 CET:
>Good evening @knz,
>can i start working on the issue "cli: missing username in connection
>URL prevents username from appearing in prompt" ? I will find the
>appropriate thread from the list of issues and continue my feedback
>there.
>
>Best regards,
>Panos.
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/cockroachdb/cockroach/pull/13379#issuecomment-279032275

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
